### ISSUE TITLE
Workflow: How to ?

### ISSUE THREAD

**Issue by theova** [InspectorMustache/base16-builder-python#8]
Hello!

Thank your work on this builder. It's amazing to produce so many themes with your tool :+1:

I'm mainting [base16-qutebrowser](https://github.com/theova/base16-qutebrowser) and would like to ask you about a good workflow with base16-builder-python.

### My folder structure
- base16-qutebrowser
  - templates
    - config.yaml 
    - default.mustache
  - themes
    - base16-3024.config.py
    - ...
    - base16-zenburn.config.py

### Desired workflow
- use `pybase16` to update all schemes from https://github.com/chriskempson/base16-schemes-source.
- use `pybase build [options]` to generate all themes by reading `base16-qutebrowser/templates/config.yaml` and `base16-qutebrowsertemplates/default.mustache` and storing them in `base16-qutebrowser/themes`.

### Current workflow
I use the following Makefile:
``` Makefile
BUILD=pybase16
TEMPLATE=qutebrowser
THEME_FOLDER=themes
OUTPUT=output

all: update build

update: 
	$(BUILD) update
build:
	$(BUILD) build -t ${TEMPLATE} -o ${OUTPUT}
	rm -r ${THEME_FOLDER}
	mv ${OUTPUT}/${TEMPLATE}/themes/ ${THEME_FOLDER}/
	rm -r ${OUTPUT}
```
#### Problem: 
`pybase update` downloads all templates to `base16-qutebrowser/templates/`, e.g. `base16-qutebrowser/templates/qutebrowser/` as well. Beside the fact, that I don't need all templates, it clones my own repo with the (old) `default.mustache` as well. This is rather confusing and the old cloned `default.mustache` is used to generate all templates.

So I would like to be able to only update the schemes (not the templates) sources and/or specify which `default.mustache` it has to use.

Is this already possible? Otherwise it would be a very cool feature!

Thank you very much!

**Comment by InspectorMustache** [comment]
Hi, and thanks for the kind words!

You can control which templates and schemes are downloaded as part of the update operation by editing the source files and then passing the -c flag. I'll try to explain a little more in detail:

The update operation basically consists of three steps:
1. Create a file called sources.yaml. This file contains URLs to [this repo](https://github.com/chriskempson/base16-schemes-source.git) and [this repo](https://github.com/chriskempson/base16-templates-source.git). These repositories each contain a list of scheme and template repositories, respectively, and are maintained by the base16 author. (You probably already know this, seeing as the qutebrowser template is in there as well.)
2. Clone the repositories in sources.yaml.
3. Go into the two downloaded repositories from the previous step and open the file `list.yaml` in each of these. This is the file that contains a list of all the scheme or template repositories. Every repository in these two files is then cloned as well.

You can skip steps one and two by providing the `-c/--custom` flag. This effectively means that you can edit the `list.yaml` files and remove those template or scheme repositories you don't need. In your case, you could just remove every repository from `sources/templates/list.yaml` or just delete the file altogether. **But update the builder first** because a missing or empty source file caused an exception to be raised and I only just fixed this.

So in short: run `update`, delete everything you don't want, delete `sources/templates/list.yaml` and run `update -c`. If I understand correctly this also takes care of the problem with the `default.yaml` file, right?

This is probably not the most intuitive way for controlling the cloning process but I think it's the most flexible. I also can't think of a much better way of excluding specific repos from the update process.

**Comment by theova** [comment]
Hi,

thank you very much for your detailed explanation.

Now I have a `Makefile` fullfilling my wishes:
```Makefile
BUILD=pybase16
TEMPLATE=qutebrowser
THEME_DIR=themes
SOURCE_DIR=sources
TEMPLATE_DIR=templates
OUTPUT=output

all: update build

update:
	$(BUILD) update
	rm -rf ${TEMPLATE_DIR}/*/
	rm -rf ${SOURCE_DIR}/${TEMPLATE_DIR}
	$(BUILD) update -c
build:
	mkdir -p ${TEMPLATE_DIR}/${TEMPLATE}/templates
	cp ${TEMPLATE_DIR}/*.* ${TEMPLATE_DIR}/${TEMPLATE}/templates
	$(BUILD) build -t ${TEMPLATE} -o ${OUTPUT}
	rm -rf ${THEME_DIR}
	mv ${OUTPUT}/${TEMPLATE}/templates/ ${THEME_DIR}/
	rm -rf ${OUTPUT}
	rm -rf ${TEMPLATE_DIR}/*/
```

Im not sure, if `pybase16 update -c` is even necessary.

> If I understand correctly this also takes care of the problem with the default.yaml file, right?

Yes :-)

**Comment by InspectorMustache** [comment]
With that procedure, `update -c` isn't necessary because you're overwriting the old qutebrowser template with the new one in the next step. But you don't actually have to move stuff around so much. The `-t` flag of the build operation accepts a path to a template folder. So you could just cd into a directory outside of your repository and do:
```
pybase16 update
pybase16 build -t /path/to/your/qutebrowser/template/repo
```

**Comment by theova** [comment]
Thanks a lot! Now I'm 100% satisfied.

Just for someone else, my `Makefile` is:
```
BUILD=pybase16
REPO=$(shell pwd)
TEMPLATE=$(shell basename ${REPO})
THEME_DIR=themes
TEMPLATE_DIR=templates
OUTPUT=output

all: update build

update:
	$(BUILD) update

build:
	$(BUILD) build -t ${REPO} -o ${OUTPUT}
	rm -rf ${THEME_DIR}
	mv ${OUTPUT}/${TEMPLATE}/themes/ ${THEME_DIR}/
	rm -rf ${OUTPUT} ${TEMPLATE_DIR}/*/
```
